### DISCUSSION TITLE
Best model is not best

### DISCUSSION THREAD

**Discussion by sunhuaiqiang** [pycaret/pycaret#2233] [Q&A]
I just write a function to compare the best model generated by compare_models() with extraTrees under identical settings for several random seeds:

`def test_bestClassifier(data,id):

    setting = setup(data = data, target = 'GROUP', preprocess=False, session_id=id, train_size=0.75, html=False, silent=True, use_gpu=False)     

    best = compare_models(exclude=exclude_list)
    model_name = get_model_name(best)
    best_model = create_model(model_name)
    best_model_tuned = tune_model(best_model, search_library='optuna')
    pred_best = predict_model(best_model_tuned)
    acc_best = check_metric(pred_best['GROUP'], pred_best['Label'], metric='Accuracy')
    auc_best = check_metric(pred_best['GROUP'], pred_best['Label'], metric='AUC')  

    et = create_model('et')  
    tuned_et = tune_model(et,search_library='optuna')
    pred_et = predict_model(tuned_et)
    acc_et = check_metric(pred_et['GROUP'], pred_et['Label'], metric='Accuracy')
    auc_et = check_metric(pred_et['GROUP'], pred_et['Label'], metric='AUC')  

    return {'et':acc_et},{'et':auc_et},{model_name:acc_best},{model_name:auc_best},model_name`

Theoretically, the performance of 'et' should not be better than the best model, However, the results show that the et outperformed the best model in most cases, in terms of both accuracy and auc. Can any expert help me to explain

**Comment by ngupta23** [comment]
The best model obtained is without any tuning so, in that sense, it is still the best model. After tuning, the best model may be different.

Also, note you do not need to recreate the best model after the compare_model call. You can directly do this.

```python
best = compare_models(exclude=exclude_list)
# model_name = get_model_name(best)  # Not needed
# best_model = create_model(model_name)  # Not needed
best_model_tuned = tune_model(best, search_library='optuna')  # Replace best_model with best which is the same.
```

**Reply by sunhuaiqiang** [reply]
> The best model obtained is without any tuning so, in that sense, it is still the best model. After tuning, the best model may be different.
> 
> Also, note you do not need to recreate the best model after the compare_model call. You can directly do this.
> 
> ```python
> best = compare_models(exclude=exclude_list)
> # model_name = get_model_name(best)  # Not needed
> # best_model = create_model(model_name)  # Not needed
> best_model_tuned = tune_model(best, search_library='optuna')  # Replace best_model with best which is the same.
> ```

Thank you very much for your reply, but I still have doubts about compare_models(). I remove tune step from both specified model and the best model from compare_models(), but still found 'et' outperformed the best model in at least half of the cases.

`
setting = setup(data = data, target = 'GROUP', preprocess=False, session_id=id, train_size=0.75, html=False, silent=True, use_gpu=False)     
best = compare_models(exclude=exclude_list)
best_model = create_model(best)
pred_best = predict_model(best_model)
acc_best = check_metric(pred_best['GROUP'], pred_best['Label'], metric='Accuracy')
auc_best = check_metric(pred_best['GROUP'], pred_best['Label'], metric='AUC')  

et = create_model('et')  
pred_et = predict_model(et)
acc_et = check_metric(pred_et['GROUP'], pred_et['Label'], metric='Accuracy')
auc_et = check_metric(pred_et['GROUP'], pred_et['Label'], metric='AUC') 
 `
